### PR TITLE
Fix Issue #119 by Raising IDNAError on non-ASCII A-label

### PR DESCRIPTION
**Pull request by OrianeK**

### CODE REVIEW

File: tests/test_idna.py [hunk 1]
```diff
@@ -261,6 +262,6 @@ def test_decode(self):
         self.assertRaises(idna.IDNAError, idna.decode, 'xn--')
         self.assertRaises(idna.IDNAError, idna.decode, b'\x8d\xd2')
         self.assertRaises(idna.IDNAError, idna.decode, b'A.A.0.a.a.A.0.a.A.A.0.a.A.0A.2.a.A.A.0.a.A.0.A.a.A0.a.a.A.0.a.fB.A.A.a.A.A.B.A.A.a.A.A.B.A.A.a.A.A.0.a.A.a.a.A.A.0.a.A.0.A.a.A0.a.a.A.0.a.fB.A.A.a.A.A.B.0A.A.a.A.A.B.A.A.a.A.A.a.A.A.B.A.A.a.A.0.a.B.A.A.a.A.B.A.a.A.A.5.a.A.0.a.Ba.A.B.A.A.a.A.0.a.Xn--B.A.A.A.a')
-
+
```
Comment (sethmlarson) [inline tests/test_idna.py:265]: Revert the changes to whitespace.

File: tests/test_idna.py [hunk 2]
```diff
@@ -246,7 +246,8 @@ def test_encode(self):
         self.assertEqual(idna.encode('a'*63), b'a'*63)
         self.assertRaises(idna.IDNAError, idna.encode, 'a'*64)
         self.assertRaises(idna.core.InvalidCodepoint, idna.encode, '*')
-
+        self.assertRaises(idna.IDNAError, idna.encode, b'\x0a\x33\x81')
+
```
Comment (sethmlarson) [inline tests/test_idna.py:250]: Revert the change to whitespace.

File: idna/core.py [hunk 3]
```diff
@@ -338,8 +338,11 @@ def uts46_remap(domain: str, std3_rules: bool = True, transitional: bool = False
 
 
 def encode(s: Union[str, bytes, bytearray], strict: bool = False, uts46: bool = False, std3_rules: bool = False, transitional: bool = False) -> bytes:
-    if isinstance(s, (bytes, bytearray)):
-        s = s.decode('ascii')
+    try:
+        if isinstance(s, (bytes, bytearray)):
+            s = s.decode('ascii')
+    except UnicodeDecodeError:
+        raise IDNAError('Invalid ASCII in A-label')
```
Comment (sethmlarson) [inline idna/core.py:345]: Not a comment about this change, but I'm confused by this call to `.decode('ascii')` here. Isn't the point of `idna.encode()` here to transform non-ASCII characters into IDNA-encoded ASCII? If so this code path almost seems like we're maintaining it for backwards compatibility reasons instead of "functionality". Should our `IDNAError` here instead hint users to pass the value as a decoded string to force users to specify the encoding?

**Review by sethmlarson — CHANGES_REQUESTED** [review]
Some nits and a more meta comment:

File: idna/core.py (hunk 3 above)
Comment (jribbens) [reply idna/core.py:345]: I agree the error message would be more helpful if it simply said that you should be passing a unicode string to the function rather than a byte string.

Comment (jribbens) [reply idna/core.py:345]: Also I would imagine that `if ... try ...` rather than `try ... if ...` would be more efficient for the vastly more common case that a unicode string is being passed.

File: tests/test_idna.py (hunk 1 above)
Comment (OrianeK) [reply tests/test_idna.py]: Modified

File: tests/test_idna.py (hunk 2 above)
Comment (OrianeK) [reply tests/test_idna.py]: Modified

File: idna/core.py (hunk 3 above)
Comment (OrianeK) [reply idna/core.py]: Agreed and modified as suggested

**Comment by OrianeK** [comment]
Thank you for your feedback. I Agree with all your suggestions and committed the suggested changes.
